### PR TITLE
Rewrite lookup index selection implementation

### LINKED ISSUE #1061: Add a scan operator and refactor the lookup statement.

### PR DESCRIPTION
**Pull request by yixinglu**
closes #1061

### CODE REVIEW

File: src/validator/LookupValidator.cpp [hunk 1]
```diff
@@ -5,219 +5,162 @@
...
-        returnCols_->emplace_back(kRank);
-        idxScanColNames_.emplace_back(kRanking);
-        outputs_.emplace_back(idxScanColNames_.back(), Value::Type::INT);
-    } else {
-        returnCols_->emplace_back(kVid);
-        idxScanColNames_.emplace_back(kVertexID);
-        outputs_.emplace_back(idxScanColNames_.back(), vidType_);
-    }
-    if (sentence->yieldClause() == nullptr) {
+    auto yieldClause = sentence()->yieldClause();
```
Comment (CPWstatic) [inline src/validator/LookupValidator.cpp:64]: It seems you forget to handle `outputs_`.

Comment (yixinglu) [reply src/validator/LookupValidator.cpp:64]: fixed. thx.

File: src/optimizer/OptimizerUtils.cpp [hunk 2]
```diff
@@ -463,14 +479,367 @@ Value OptimizerUtils::normalizeValue(const meta::cpp2::ColumnDef& col, const Val
...
+            if (hints[i].score > rhs.hints[i].score) {
+                return false;
+            }
+        }
+        return hints.size() < rhs.hints.size();
+    }
+};
+
+Status checkValue(const ColumnDef& field, BVO bvo, Value* value) {
+    if (*value == Value()) {
```
[the remark this answers is not in the document]
Comment (yixinglu) [reply src/optimizer/OptimizerUtils.cpp:600]: fixed.

File: src/optimizer/OptimizerUtils.cpp [hunk 3]
```diff
@@ -463,14 +479,367 @@ Value OptimizerUtils::normalizeValue(const meta::cpp2::ColumnDef& col, const Val
...
+        case Expression::Kind::kRelLE: {
+            // if c1 <= int(5) , the range pair should be (min, 6)
+            // if c1 < int(5), the range pair should be (min, 5)
+            auto v = OptimizerUtils::boundValue(col, BoundValueOperator::GREATER_THAN, val);
+            if (v == Value::kNullBadType) {
+                LOG(ERROR) << "Get bound value error. field : " << col.get_name();
+                return Status::Error("Get bound value error. field : %s", col.get_name().c_str());
+            }
+            // where c <= 1 and c <= 2 , 1 should be valid.
+            if (end == Value()) {
```
[the remark this answers is not in the document]
Comment (yixinglu) [reply src/optimizer/OptimizerUtils.cpp:508]: fixed.

File: src/optimizer/OptimizerUtils.cpp [hunk 4]
```diff
@@ -463,14 +479,367 @@ Value OptimizerUtils::normalizeValue(const meta::cpp2::ColumnDef& col, const Val
...
+namespace {
+
+// {2, 1, 0} >
+// {2, 1} >
+// {2, 0, 1} >
+// {2, 0} >
+// {2} >
+// {1, 2} >
+// {1, 1} >
+// {1}
```
[the remark this answers is not in the document]
Comment (yixinglu) [reply src/optimizer/OptimizerUtils.cpp]: add more meaningful comments

**Comment by Shylock-Hg** [comment]
Really big change, could provide an overview explanation?

**Comment by yixinglu** [comment]
> Really big change, could provide an overview explanation?

of course, [here](https://confluence.nebula-graph.io/pages/viewpage.action?pageId=10718164) is some thoughts when implementing this PR. I'll refine this document later.

File: tests/tck/features/lookup/LookUp.IntVid.feature [hunk 5]
```diff
@@ -1,104 +1,5 @@
 Feature: LookUpTest_Vid_Int
 
-  Scenario: LookupTest IntVid SimpleVertex
```
Comment (Shylock-Hg) [inline tests/tck/features/lookup/LookUp.IntVid.feature:3]: Why delete these cases?

Comment (yixinglu) [reply tests/tck/features/lookup/LookUp.IntVid.feature:3]: Share these scenarios in another feature file for spaces with different vid types.

File: src/optimizer/OptimizerUtils.cpp [hunk 6]
```diff
@@ -463,14 +482,399 @@ Value OptimizerUtils::normalizeValue(const meta::cpp2::ColumnDef& col, const Val
...
+            if (begin.empty()) {
+                begin = v;
+            } else {
+                begin = v < begin ? begin : v;
+            }
+            break;
+        }
+        default: {
+            // TODO(yee): Semantic error
+            return Status::Error("Invalid expression kind.");
```
Comment (Shylock-Hg) [inline src/optimizer/OptimizerUtils.cpp:554]: Why don't change it to SemanticError directly.

Comment (yixinglu) [reply src/optimizer/OptimizerUtils.cpp:554]: It's in optimizer module. It's better to do some checks in validator.

**Review by CPWstatic — APPROVED** [review]

File: src/optimizer/OptimizerUtils.cpp [hunk 7]
```diff
@@ -463,14 +482,399 @@ Value OptimizerUtils::normalizeValue(const meta::cpp2::ColumnDef& col, const Val
...
+    for (auto& field : index.get_fields()) {
+        ScoredColumnHint hint;
+        std::vector<Expression*> operands;
+        if (!getIndexColumnHintInExpr(field, expr, &hint, &operands)) {
+            break;
+        }
+        result.hints.emplace_back(std::move(hint));
+        for (auto op : operands) {
+            usedOperands.insert(op);
+        }
```
Comment (bright-starry-sky) [inline src/optimizer/OptimizerUtils.cpp:772]: Could give some comments? If the first column is EQ logic expression, the error will be returned via line 775?

Comment (yixinglu) [reply src/optimizer/OptimizerUtils.cpp:772]: EQ expression is relational expression, this case will be handled by `selectRelExprIndex` function.

**Review by bright-starry-sky — APPROVED** [review]

**Pull request merged**